src/vbk/p2p_sync.cpp

```diff
@@ -1,57 +1,145 @@
+std::map<NodeId, PopDataNodeState> mapPopDataNodeState;
+
+template <typename PopDataType>
+bool processGetPopData(CNode* node, CConnman* connman, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    std::vector<std::vector<uint8_t>> requested_data;
+    vRecv >> requested_data;
+
+    if (requested_data.size() > MAX_POP_DATA_SENDING_AMOUNT) {
+        Misbehaving(node->GetId(), 20, strprintf("message getdata size() = %u", requested_data.size()));
+        return false;
+    }
+
+    auto& known_set = mapPopDataNodeState[node->GetId()].getSet<PopDataType>();
+
+    const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+    for (const auto& data_hash : requested_data) {
+        const PopDataType* data = pop_mempool.get<PopDataType>(data_hash);
+        known_set.insert(data_hash);
+        if (data != nullptr) {
+            connman->PushMessage(node, msgMaker.Make(PopDataType::name(), *data));
+        }
+    }
+
+    return true;
+}
+
+template <typename PopDataType>
+bool processOfferPopData(CNode* node, CConnman* connman, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    LogPrint(BCLog::NET, "received offered pop data: %s, bytes size: %d\n", PopDataType::name(), vRecv.size());
+    std::vector<std::vector<uint8_t>> offered_data;
+    vRecv >> offered_data;
+
+    if (offered_data.size() > MAX_POP_DATA_SENDING_AMOUNT) {
+        Misbehaving(node->GetId(), 20, strprintf("message getdata size() = %u", offered_data.size()));
+        return false;
+    }
+
+    auto& known_set = mapPopDataNodeState[node->GetId()].getSet<PopDataType>();
+
+    std::vector<std::vector<uint8_t>> requested_data;
+    const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+    for (const auto& data_hash : offered_data) {
+        if (!pop_mempool.get<PopDataType>(data_hash)) {
+            requested_data.push_back(data_hash);
+            known_set.insert(data_hash);
+        }
+
+        if (requested_data.size() == MAX_POP_DATA_SENDING_AMOUNT) {
+            connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+            requested_data.clear();
+        }
+    }
+
+    if (!requested_data.empty()) {
+        connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+    }
+
+    return true;
+}
+
+template <typename PopDataType>
+bool processPopData(CNode* node, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    LogPrint(BCLog::NET, "received pop data: %s, bytes size: %d", PopDataType::name(), vRecv.size());
+    PopDataType data;
+    vRecv >> data;
+
+    altintegration::ValidationState state;
+    if (!pop_mempool.submit(data, state)) {
+        LogPrint(BCLog::NET, "VeriBlock-PoP: %s ", state.GetPath());
```

Comment:
add meaningful message
```suggestion
        LogPrint(BCLog::NET, "peer %d sent statelessly invalid pop data: %s ", <PEERNUMBER>, state.GetPath());
```

Reply:
fixed